### ISSUE TITLE
 `/api/v1/schemas/{id}/validate` is deleted but still being used

### ISSUE THREAD

**Issue by phil-dotchon** [CiscoDevNet/ansible-mso#429]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The API endpoint `/api/v1/schemas/{id}/validate` was deleted in NDO 4.0.1
https://developer.cisco.com/docs/nexus-dashboard-orchestrator/4-1-2/api-changelog/#release-401-2022-08-02

Yet it seems this is being used in the Ansible modules. I'm getting failures when deploying PBR enabled contracts, which don't happen when I deploy inside the GUI, and I'm wondering if my code is creating a bad config or if the Ansible module should not be using this validate endpoint?
https://github.com/CiscoDevNet/ansible-mso/blob/ebf25f719a6baacfaedf843e5f372a836f310fdc/plugins/modules/ndo_schema_template_deploy.py#L19

If I deploy in the GUI it works OK

```
    - name: Deploy configuration to ACI Fabrics
      cisco.mso.ndo_schema_template_deploy:
        schema: CI8777772_phd_001
        template: CommonTemplate_001
        output_level: debug
      register: aci_config_via_mso
      ignore_errors: true

    - name: debug
      ansible.builtin.debug:
        var: aci_config_via_mso

TASK [debug] ***************************************************************************************************************************
ok: [e_us1_mso] => {
    "aci_config_via_mso": {
        "changed": false,
        "current": {},
        "data": null,
        "failed": true,
        "httpapi_logs": [
            [
                "info",
                "send_request() - send_request method called"
            ],
            [
                "info",
                "send_request() - connection.send(/mso/api/v1/schemas/list-identity, null, GET, {'Content-Type': 'application/json'})"
            ],
            [
                "vvv",
                "ESTABLISH HTTP(S) CONNECTFOR USER: svc-e-us1-paas-ta TO https://e-us1-ndo.fis.lab:443"
            ],
            [
                "info",
                "login() - login method called for e-us1-ndo.fis.lab"
            ],
            [
                "info",
                "login() - previous auth not found sending login POST to e-us1-ndo.fis.lab"
            ],
            [
                "info",
                "login() - connection.send(/login, LOGIN_PAYLOAD_NOT_SHOWN, POST, {'Content-Type': 'application/json'})"
            ],
            [
                "info",
                "send_request() - send_request method called"
            ],
            [
                "info",
                "send_request() - connection.send(/mso/api/v1/schemas/665d93a6a14fd9c917bf355c/validate, null, GET, {'Content-Type': 'application/json'})"
            ]
        ],
        "info": {
            "body": {
                "code": 400,
                "info": null,
                "message": "Err Missing Ref: /templates//contracts/CI9380650"
            },
            "cache-control": "no-cache, must-revalidate, proxy-revalidate, max-age=0",
            "connection": "close",
            "content-language": "en",
            "content-length": "85",
            "content-security-policy": "default-src 'self'; script-src 'self' 'unsafe-eval' 'unsafe-inline'; style-src-elem 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; base-uri 'self'; frame-ancestors 'self'; font-src 'self' data:; img-src 'self' data: alln-extcloud-storage.cisco.com; navigate-to *; worker-src 'self' blob:; connect-src 'self' wss:;",
            "content-type": "application/json",
            "error": {
                "code": -1,
                "message": {
                    "code": 400,
                    "info": null,
                    "message": "Err Missing Ref: /templates//contracts/CI9380650"
                }
            },
            "method": "GET",
            "msg": "OK (85 bytes)",
            "referrer-policy": "strict-origin-when-cross-origin",
            "server": "nginx",
            "status": 400,
            "strict-transport-security": "max-age=31536000; includeSubDomains",
            "url": "https://x.x.x:443/mso/api/v1/schemas/665d93a6a14fd9c917bf355c/validate",
            "x-content-type-options": "nosniff",
            "x-frame-options": "SAMEORIGIN",
            "x-ratelimit-limit": "-1",
            "x-ratelimit-remaining": "-1",
            "x-ratelimit-reset": "1559582945",
            "x-xss-protection": "1; mode=block"
        },
        "method": "GET",
        "msg": "MSO Error 400: Err Missing Ref: /templates//contracts/CI9380650",
        "payload": {
            "code": 400,
            "info": null,
            "message": "Err Missing Ref: /templates//contracts/CI9380650"
        },
        "response": "OK (85 bytes)",
        "socket": "/home/lc5648734/.ansible/pc/d536010bd7",
        "status": 400,
        "url": "https://x.x.x:443/mso/api/v1/schemas/665d93a6a14fd9c917bf355c/validate"
    }
}
```

If I GET `/mso/api/v1/schemas/665d93a6a14fd9c917bf355c/validate` (as above) in Postman, the response is different:

`{"code":400,"message":"Err Missing Ref: /schemas/665d93a6a14fd9c917bf355c/templates/CommonTemplate_001/serviceGraphs/","info":null}`

### Affected Module Name(s):

* ndo_schema_template_deploy.py

### MSO version and MSO Platform

* V 4.2.3e and ND-based.

### APIC version and APIC Platform for Site Level Resources

* V 5.3.2b and on-prem/cloud-aws/cloud-azure/all.

### Collection versions

* cisco.mso 2.6.0

### Output/ Error message

```
    - name: Deploy configuration to ACI Fabrics
      cisco.mso.ndo_schema_template_deploy:
        schema: CI8777772_phd_001
        template: CommonTemplate_001
        output_level: debug
      register: aci_config_via_mso
      ignore_errors: true

    - name: debug
      ansible.builtin.debug:
        var: aci_config_via_mso
```

### Expected Behavior

* Deploy completes successfully 

### Actual Behavior

* Deploy calls /validate which fails

### Playbook tasks to Reproduce

*

### References

* #0000

**Comment by phil-dotchon** [comment]
I think I found the root of my deployment failures, I was apparently sending patch content that is not required/valid in this newer version of NDO. That aside, happy to close this if you think there is no concern with the Ansible modules using the /validate endpoint. It does appear that the GUI doesn't do that

**Comment by akinross** [comment]
Hi @phil-dotchon, 

Thank you for raising the issue and good to see your problem is resolved. I will keep the issue open for now, because I want to follow up internally on the validation endpoint. I will provide updates on my findings, this will likely take a bit of time.

**Comment by akinross** [comment]
I did a internal check and we will be able to use it as an exception in our ansible collection, it is not used in the UI because the validation engine is triggered upon save. Since from collection point of view there are a few exception in which we had to bypass the validation engine it is needed to ensure that schemas that are not validated are not deployed via the module.

Hope this answers your questions, will close the issue for now.
